worker/provisioner/provisioner_test.go

```diff
@@ -1894,10 +1894,22 @@ func (s *ProvisionerSuite) TestProvisioningMachinesDerivedAZ(c *gc.C) {
 	mSucceed := machines[2:]
 
 	s.checkStartInstances(c, mSucceed)
-	c.Assert(e.retryCount[mFail[1].Id()], gc.Equals, 3)
 	c.Assert(e.retryCount[mSucceed[0].Id()], gc.Equals, 1)
 	c.Assert(e.retryCount[mSucceed[2].Id()], gc.Equals, 1)
 
+	// This synchronisation addresses a potential race condition.
+	// It can happen that upon successful return from checkStartInstances
+	// The machine(s) arranged for provisioning failure have not yet been
+	// retried the specified number of times; so we wait.
+	id := mFail[1].Id()
+	for e.retryCount[id] < 3 {
+		select {
+		case <-time.After(coretesting.ShortWait):
```

Comment:
I don't think this does what you want. What you want is:
after := time.After(coretesting.LongWait)
for e.RetryCount[id] < 3 {
  select {
    case <-after:
      c.Fatalf()
    default:
  }
}

Also, this is a case where the wait should probably be LongWait. Which is what you use for "something should have happened, but we haven't noticed yet". That way we don't fail when the machine we are running on is slower than expected.
ShortWait is more for "something should not have happened, wait this long to give it a chance to have happened".


In this code, the time.After is restarted every time through the loop, so it will never fail.